File: variable/test/cumulative_test.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
#include <gtest/gtest.h>

#include "scipp/core/except.h"
#include "scipp/variable/bins.h"
#include "scipp/variable/cumulative.h"

using namespace scipp;

TEST(CumulativeTest, exclusive_scan) {
  const auto var = makeVariable<int64_t>(Dims{Dim::X, Dim::Y}, Shape{2, 3},
                                         Values{1, 2, 3, 4, 5, 6});
  EXPECT_EQ(exclusive_scan(var, Dim::X),
            makeVariable<int64_t>(var.dims(), Values{0, 0, 0, 1, 2, 3}));
  EXPECT_EQ(exclusive_scan(var, Dim::Y),
            makeVariable<int64_t>(var.dims(), Values{0, 1, 3, 0, 4, 9}));
}

TEST(CumulativeTest, exclusive_scan_bins) {
  const auto indices =
      makeVariable<scipp::index_pair>(Values{scipp::index_pair{0, 3}});
  const auto buffer =
      makeVariable<int64_t>(Dims{Dim::Row}, Shape{3}, Values{1, 2, 3});
  const auto var = make_bins(indices, Dim::Row, buffer);
  EXPECT_EQ(exclusive_scan_bins(var),
            make_bins(indices, Dim::Row,
                      makeVariable<int64_t>(buffer.dims(), Values{0, 1, 3})));
}

File: tests/cumulative_test.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
import scipp as sc


def test_cumsum() -> None:
    var = sc.arange('xx', 4, unit='m')
    assert sc.identical(
        sc.cumsum(var), sc.array(dims=['xx'], values=[0, 1, 3, 6], unit='m')
    )
    assert sc.identical(
        sc.cumsum(var, dim='xx'), sc.array(dims=['xx'], values=[0, 1, 3, 6], unit='m')
    )
    assert sc.identical(
        sc.cumsum(var, mode='exclusive'),
        sc.array(dims=['xx'], values=[0, 0, 1, 3], unit='m'),
    )
    assert sc.identical(
        sc.cumsum(var, dim='xx', mode='exclusive'),
        sc.array(dims=['xx'], values=[0, 0, 1, 3], unit='m'),
    )
